server/src/main/java/org/elasticsearch/indices/recovery/RecoverySourceHandler.java

```
@@ -634,73 +635,82 @@ public String toString() {
-                    } else {
-                        throw e;
                     }
+                } catch (IOException e) {
+                    handleErrorOnSendFiles(store, md, e);
                 }
             }
         } finally {
-            store.decRef();
+            try {
+                sendingTickets.acquire(maxConcurrentFileChunks);
+                assert requestSeqIdTracker.getCheckpoint() == seqIdGenerator : "not all requests completed; " +
+                    "completed_requests=" + requestSeqIdTracker.getCheckpoint() + " total_requests=" + seqIdGenerator;
+            } finally {
+                if (error.get() != null) {
+                    handleErrorOnSendFiles(store, error.get().v1(), error.get().v2());
+                }
+            }
+        }
+    }
+
+    private void handleErrorOnSendFiles(Store store, StoreFileMetaData md, Exception e) throws IOException {
+        final IOException corruptIndexException;
+        if ((corruptIndexException = ExceptionsHelper.unwrapCorruption(e)) != null) {
+            if (store.checkIntegrityNoException(md) == false) { // we are corrupted on the primary -- fail!
+                logger.warn("{} Corrupted file detected {} checksum mismatch", shardId, md);
+                failEngine(corruptIndexException);
+                throw corruptIndexException;
+            } else { // corruption has happened on the way to replica
+                RemoteTransportException exception = new RemoteTransportException(
+                    "File corruption occurred on recovery but checksums are ok", null);
+                exception.addSuppressed(e);
+                logger.warn(() -> new ParameterizedMessage("{} Remote file corruption on node {}, recovering {}. local checksum OK",
+                    shardId, request.targetNode(), md), corruptIndexException);
+                throw exception;
+            }
+        } else {
+            throw ExceptionsHelper.convertToRuntime(e);
```

why change this? Let's throw the original exception like it used to be before